net/proxymux/mux.go

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) 2021 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package proxymux splits a net.Listener in two, routing SOCKS5
+// connections to one and HTTP requests to the other.
+//
+// It allows for hosting both a SOCKS5 proxy and an HTTP proxy on the
+// same listener.
+package proxymux
+
+import (
+	"io"
+	"net"
+	"sync"
+)
+
+// Mux accepts connections on l and passes connections through to
+// either socksListener or httpListener, depending the first byte sent
+// by the client.
+func Mux(l net.Listener) (socksListener, httpListener net.Listener) {
```

Comment:
s/l/ln/ which Go style has moved towards (avoiding the variable `l` which looks like `1` in many people's fonts)

Reply:
rename "Mux" to "SplitSOCKSAndHTTP" or something to keep this package generic for the future for other muxing or configurability? then the name also describes the order of the result params so caller can't transpose so easily.

(this is almost the tcpproxy thing we worked on before if you squint)

Reply:
Renamed, adjusted varnames